crypto/thread/api.c

```diff
@@ -0,0 +1,169 @@
+{
+    return 1;
+}
+
+int CRYPTO_THREAD_cap(OSSL_LIB_CTX *ctx, int max_threads)
+{
+    return 0;
+}
+
+int CRYPTO_THREAD_spawn_worker(OSSL_LIB_CTX *ctx, CRYPTO_WORKER_CALLBACK cb)
+{
+    return 0;
+}
+
+#else
+
+# include <internal/thread.h>
+# include "openssl_threads.h"
+# include "worker.h"
+
+static ossl_inline int thread_count_active(OSSL_LIB_CTX *ctx);
+
+int CRYPTO_THREAD_enabled(OSSL_LIB_CTX *ctx)
+{
+    int enabled;
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    tdata = OSSL_LIB_CTX_GET_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    CRYPTO_MUTEX_lock(tdata->lock);
+    enabled = tdata->enabled;
+    CRYPTO_MUTEX_unlock(tdata->lock);
+
+    return enabled;
+}
+
+int CRYPTO_THREAD_enable(OSSL_LIB_CTX *ctx, int max_threads)
+{
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    tdata = OSSL_LIB_CTX_GET_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    CRYPTO_MUTEX_lock(tdata->lock);
+    tdata->workers.internal.cap = max_threads;
+    tdata->enabled = 1;
+    CRYPTO_MUTEX_unlock(tdata->lock);
+
+    return 1;
+}
+
+int CRYPTO_THREAD_disable(OSSL_LIB_CTX *ctx)
+{
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    tdata = OSSL_LIB_CTX_GET_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    if (tdata->enabled == 0)
+        return 1;
+
+    /*
+     * Cannot disable threads if there's a thread running or when there's a
+     * queued task that can still be processed.
+     */
+    CRYPTO_MUTEX_lock(tdata->lock);
+    if (thread_count_active(ctx) > 0) {
+        CRYPTO_MUTEX_unlock(tdata->lock);
+        return 0;
+    }
+    if (list_size(&tdata->tasks.queue) > 0
+        && (list_size(&tdata->workers.external.available) > 0
+                  || tdata->workers.internal.cap != 0)) {
+            CRYPTO_MUTEX_unlock(tdata->lock);
+            return 0;
+    }
+
+    tdata->workers.internal.cap = 0;
+    tdata->enabled = 0;
+    CRYPTO_MUTEX_unlock(tdata->lock);
+
+    if (worker_terminate_all(ctx) == 0)
+        return 0;
+    if (worker_clean_all(ctx) == 0)
+        return 0;
+    if (crypto_task_clean_all(ctx) == 0)
+        return 0;
+
+    return 1;
+}
+
+int CRYPTO_THREAD_cap(OSSL_LIB_CTX *ctx, int max_threads)
+{
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    tdata = OSSL_LIB_CTX_GET_THREADS(ctx);
+    if (tdata == NULL)
+        return 0;
+
+    if (tdata->enabled == 0)
+        return 0;
+
+    CRYPTO_MUTEX_lock(tdata->lock);
+    tdata->workers.internal.cap = max_threads;
+    CRYPTO_MUTEX_unlock(tdata->lock);
+    return 1;
+}
+
+int CRYPTO_THREAD_spawn_worker(OSSL_LIB_CTX *ctx, CRYPTO_WORKER_CALLBACK cb)
+{
+    int retval;
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    tdata = OSSL_LIB_CTX_GET_THREADS(ctx);
+    if (tdata == NULL || tdata->enabled == 0)
+        return 0;
+
+    CRYPTO_MUTEX_lock(tdata->lock);
+    retval = thread_spawn_worker_task(ctx, cb, NULL);
+    CRYPTO_MUTEX_unlock(tdata->lock);
+
+    return retval;
+}
+
+static ossl_inline int thread_count_active(OSSL_LIB_CTX *ctx)
```

Comment:
Assume tdata is already locked. Worthy of a comment.

Reply:
Would you object to a macro instead?
```
#define CRYPTO_MUTEX_assert_locked(X) // not implemented
CRYPTO_MUTEX_assert_locked(tdata->lock)
```
It'd be standard all across the code and easy to find/extend in the future.